lib/cli/src/Cardano/CLI.hs

```diff
@@ -298,6 +319,147 @@ runCli = join . customExecParser preferences
+            = penult ++ " or " ++ ult
+         formatEnglishEnumerationRev (ult:penult:revBeginning)
+            = intercalate ", " (reverse revBeginning)
+                ++ ", "
+                ++ penult
+                ++ " or "
+                ++ ult
+         formatEnglishEnumerationRev xs = intercalate ", " (reverse xs)
+
+parseSeedWords :: Parser [Text]
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Lay thine eyes upon the type parameters and see that there are none.
+data CLIKeyScheme = CLIKeyScheme
+    { wordsToSeed :: [Text] -> Either String (Passphrase "seed")
+    , allowedWordLengths :: [Int]
+    , seedToRootKey :: Passphrase "seed" -> XPrv
+    }
+
+byronCliKeyScheme :: ByronWalletStyle -> CLIKeyScheme
+byronCliKeyScheme = \case
+    Random -> mkScheme (Proxy @'Random)
+            (Byron.getKey . flip Byron.generateKeyFromSeed pass)
+    Icarus ->
+        mkScheme (Proxy @'Icarus)
+            icarusStyleKeyToSeed
+    Trezor ->
+        mkScheme (Proxy @'Trezor)
+            icarusStyleKeyToSeed
+    Ledger ->
+        mkScheme (Proxy @'Ledger)
+            icarusStyleKeyToSeed
+  where
+    mkScheme
+        :: forall (s :: ByronWalletStyle).
+            ( FromMnemonic (AllowedMnemonics s) "seed"
+            , ManyNatVal (AllowedMnemonics s)
+            )
+        => Proxy s
+        -> (Passphrase "seed" -> XPrv)
+        -> CLIKeyScheme
+
+    -- Construct a KeyScheme using the type families of the promoted type
+    -- argument.
+    mkScheme _ = CLIKeyScheme
+            (left getFromMnemonicError . fromMnemonic @(AllowedMnemonics s) @"seed")
+            (map fromIntegral (manyNatVal $ Proxy @(AllowedMnemonics s)))
+
+    icarusStyleKeyToSeed =
+            (Icarus.getKey . flip Icarus.generateKeyFromSeed pass)
+
+    -- We don't use passwords to encrypt the keys here.
+    pass = mempty
+
+cmdRootKey :: Mod CommandFields (IO ())
+cmdRootKey =
+    command "root" $ info (helper <*> cmd) $ mempty
+        <> progDesc "Extract root xprv as hex\n\
+                    \ (64 bytes private key + 32 bytes chain code)"
+  where
+    cmd = do
+        keyType <- parseKeyType
+        ws <- parseSeedWords
+        return $ do
+            scheme <- byronCliKeyScheme <$> keyType
+
+            seed <- toIOErr $ wordsToSeed scheme ws
+            let xprv = seedToRootKey scheme seed
+            TIO.putStrLn $ keyToHexText xprv
+
+    toIOErr :: Either String a -> IO a
+    toIOErr (Right a) = return a
+    toIOErr (Left e) = hPutStrLn stderr e >> exitFailure
+
+
+-- | Machinery to demote type level [Nat] to a normal [Integer] list
+--
+-- We need it in this module to get the result of AllowedMnemonics as a value.
+class ManyNatVal (xs :: [Nat]) where
+  manyNatVal :: proxy xs -> [Integer]
+
+instance ManyNatVal '[] where
+  manyNatVal _ = []
+
+instance (KnownNat a, ManyNatVal as) => ManyNatVal (a ': as) where
+  manyNatVal _ =
+    natVal (Proxy :: Proxy a) : manyNatVal (Proxy :: Proxy as)
```

Comment:
This belongs to `Cardano.Wallet.Api.Types` next to `AllowedMnemonics`, don't you think?

Reply:
6c5ec2116d98396351514ca03250a6903ae73c3e